HIVE-24945: Support vectorization for lead/lag functions

abstractdog opened the pull request:
### What changes were proposed in this pull request?
Implement vectorized evaluators for lead and lag functions.


### Why are the changes needed?
For better vectorization coverage (performance).


### Does this PR introduce _any_ user-facing change?
No.


### How was this patch tested?
Performance tested on cluster, added new qtests + old qtests got vectorized without query result change.

abstractdog commented:
@ramesh0201: is there a chance that you can review this too? (much simpler than HIVE-24761 :) )

ramesh0201 commented on ql/src/java/org/apache/hadoop/hive/ql/plan/VectorPTFDesc.java:
```diff
@@ -172,8 +179,8 @@ public VectorPTFDesc() {
   public static VectorPTFEvaluatorBase getEvaluator(SupportedFunctionType functionType,
-      boolean isDistinct, WindowFrameDef windowFrameDef, Type columnVectorType,
-      VectorExpression inputVectorExpression, int outputColumnNum) {
+      List<ExprNodeDesc> evaluatorParameters, boolean isDistinct, WindowFrameDef windowFrameDef,
```
I think we dont need the evaluatorParameters argument passed to this function

abstractdog replied:
right, I'm removing this, we can put back later if needed

ramesh0201 commented on ql/src/java/org/apache/hadoop/hive/ql/exec/vector/ptf/VectorPTFGroupBatches.java:
```diff
@@ -858,6 +858,15 @@ private void runEvaluatorForRow(int evaluatorIndex, VectorPTFEvaluatorBase evalu
+       * care of them here.
+       */
+      RowPositionInBatch rp = getPosition(rowNum);
+      evaluator.evaluateInputExpr(bufferedBatches.get(rp.batchIndex));
```
Can this cause slowness/overhead on other optimized calculated PTF evaluators which doesn't need this calculation. Is this only need for lead/lag evaluators?

abstractdog replied:
I need to rethink this...I can see two problems:

1. sum/avg runs optimized calculation but still ends up calling evaluateGroupBatch which takes care of this part, so not needed (so I would say this is for lead/lag only), so maybe I have to create another method reflecting this property of an evaluator
2. this code evaluates input expression for the same batch which contains the current row, but only for that one, which can be still problematic if a lead function points to a later record, which is present in the next batch for instance...

ramesh0201 commented:
@abstractdog  Looks great except the two comments.

ramesh0201 commented:
Also probably need to rebase to latest branch and update the q.out file as it doesn't seem to contain the deleted rows query patch

abstractdog commented:
@ramesh0201 : rebased and fixed patch passed precommit testing, PR changes are in https://github.com/apache/hive/pull/2278/commits/81592ac9909d1e2875dfdcb2e2e86f1f597062b3

ramesh0201 commented:
+1. LGTM.
